File: backend/app/models/community.py
from __future__ import annotations

from datetime import date

from sqlalchemy import Boolean, Date, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base, PublishableMixin, SoftDeleteMixin, TimestampMixin


class CommunityProfile(Base, TimestampMixin):
    """Singleton row describing Velletri.dev as an initiative."""

    __tablename__ = "community_profile"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(120), nullable=False, default="Velletri.dev")
    role_en: Mapped[str] = mapped_column(String(120), nullable=False)
    role_it: Mapped[str] = mapped_column(String(120), nullable=False)
    mission_en: Mapped[str] = mapped_column(Text, nullable=False)
    mission_it: Mapped[str] = mapped_column(Text, nullable=False)
    description_en: Mapped[str] = mapped_column(Text, nullable=False)
    description_it: Mapped[str] = mapped_column(Text, nullable=False)
    vision_en: Mapped[str] = mapped_column(Text, nullable=False)
    vision_it: Mapped[str] = mapped_column(Text, nullable=False)
    collaboration_en: Mapped[str] = mapped_column(Text, nullable=False)
    collaboration_it: Mapped[str] = mapped_column(Text, nullable=False)
    founded_year: Mapped[int] = mapped_column(Integer, nullable=False)
    website_url: Mapped[str | None] = mapped_column(String(255), nullable=True)


class CommunityActivity(TimestampMixin, PublishableMixin, SoftDeleteMixin, Base):
    __tablename__ = "community_activity"

    id: Mapped[int] = mapped_column(primary_key=True)
    slug: Mapped[str] = mapped_column(String(140), unique=True, index=True, nullable=False)
    title_en: Mapped[str] = mapped_column(String(200), nullable=False)
    title_it: Mapped[str] = mapped_column(String(200), nullable=False)
    description_en: Mapped[str] = mapped_column(Text, nullable=False)
    description_it: Mapped[str] = mapped_column(Text, nullable=False)
    activity_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    activity_type: Mapped[str] = mapped_column(String(40), nullable=False, default="meetup")
    url: Mapped[str | None] = mapped_column(String(255), nullable=True)
    logo_media_url: Mapped[str | None] = mapped_column(String(255), nullable=True)
    is_featured: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    internal_notes: Mapped[str | None] = mapped_column(Text, nullable=True)

File: frontend/src/admin/types/community.ts
export interface CommunityActivityListItem {
  id: number;
  slug: string;
  title_en: string;
  activity_type: string;
  sort_order: number;
  publication_status: "DRAFT" | "SCHEDULED" | "PUBLISHED" | "ARCHIVED";
  deleted_at: string | null;
}

export interface CommunityActivityAdminDetail {
  id: number;
  slug: string;
  title_en: string;
  title_it: string;
  description_en: string;
  description_it: string;
  activity_date: string | null;
  activity_type: string;
  url: string | null;
  logo_media_url: string | null;
  is_featured: boolean;
  sort_order: number;
  internal_notes: string | null;
  publication_status: "DRAFT" | "SCHEDULED" | "PUBLISHED" | "ARCHIVED";
  publish_at: string | null;
  unpublish_at: string | null;
  published_at: string | null;
  deleted_at: string | null;
  created_at: string;
  updated_at: string;
}

export type CommunityActivityWritePayload = Omit<
  CommunityActivityAdminDetail,
  | "id"
  | "publication_status"
  | "publish_at"
  | "unpublish_at"
  | "published_at"
  | "deleted_at"
  | "created_at"
  | "updated_at"
>;

export interface CommunityActivityRevisionSummary {
  id: number;
  schema_version: number;
  snapshot: Record<string, unknown>;
  created_at: string;
  created_by_email: string | null;
}

export interface CommunityProfileAdmin {
  id: number;
  name: string;
  role_en: string;
  role_it: string;
  mission_en: string;
  mission_it: string;
  description_en: string;
  description_it: string;
  vision_en: string;
  vision_it: string;
  collaboration_en: string;
  collaboration_it: string;
  founded_year: number;
  website_url: string | null;
  created_at: string;
  updated_at: string;
}

export type CommunityProfileWritePayload = Omit<
  CommunityProfileAdmin,
  "id" | "created_at" | "updated_at"
>;
